### PR TITLE
973 front running prevention

### PR DESCRIPTION
**Pull request by dtebbs**
- [x] multiple Submissions per SubmissionIdx
- [x] simplify how `upa aggregator submit-aggregated-proof` works
- [x] optimize the Submission[] on-chain
- [x] optimize calldata for `verifyAggregatedProofs`

### CODE REVIEW

**Review by SupremoUGH — CHANGES_REQUESTED** [review]
In general, not too happy with the added complexity (and costs) given the problem we're solving. 

The contract in general looks more vulnerable now, and I'm not convinced that this is 100% safe. Maybe we can separate the PR into two to make things clearer:
1) Includes the front-running protection
2) Includes the optimizations, the new structs and `verifyAggregatedProof` reorg. 

I also expect the necessary changes in the worker to be quite non-trivial...

**Review by SupremoUGH — CHANGES_REQUESTED** [review]

**Review by SupremoUGH — APPROVED** [review]

File: spec/protocol/upa-1.2.0.md [hunk 1]
```diff
@@ -55,7 +57,9 @@ The proof and public input data is not stored on-chain. The aggregator monitors
...
 
-Once a submission is verified by the UPA contract, its submission id is marked as verified. Applications can confirm that an individual proof id is verified by providing a `ProofReference`, which is a Merkle proof that the proof id was included in a verified submission. Note that for proofs in a multi-proof submission with $\sid$, the contract does not mark the proof as verified until the entire submission has been verified.
+Once a submission is verified by the UPA contract, its submission id is marked as verified.  Applications can now query the UPA with the given submission Id and check that it has been verified.  Internall, the UPA contract only marks the submission with the given submission id and *duplicate submission index* as verified, but when applications query with a submission Id, the UPA answers in the affirmative if ANY submission with the given Submission Id is marked as verified.  In this sense, *duplicate submission index* is hidden from applications.
```
Comment (dysfunctor) [inline spec/protocol/upa-1.2.0.md]:
```suggestion
Once a submission is verified by the UPA contract, its submission id is marked as verified.  Applications can now query the UPA with the given submission Id and check that it has been verified.  Internally, the UPA contract only marks the submission with the given submission id and *duplicate submission index* as verified, but when applications query with a submission Id, the UPA answers in the affirmative if ANY submission with the given Submission Id is marked as verified.  In this sense, *duplicate submission index* is hidden from applications.
```

File: upa/contracts/UpaVerifier.sol [hunk 2]
```diff
@@ -583,7 +605,7 @@ contract UpaVerifier is
 
         bytes32 submissionId = UpaLib.computeSubmissionId(proofId);
 
-        uint64 submissionIdx = getSubmissionIdx(submissionId);
+        uint64 submissionIdx = getSubmissionIdx(submissionId, 0);
```
Comment (dysfunctor) [inline upa/contracts/UpaVerifier.sol]: I think this `isProofVerified` needs to be updated to also check the later dupSubmissionIdxs.

Can it just call `isSubmissionVerified(submissionId)`? I don't remember if there was a reason I didn't implement that way.

File: upa/contracts/UpaVerifier.sol [hunk 3]
```diff
@@ -666,46 +688,62 @@ contract UpaVerifier is
...
 
-        // On-chain submission found, check if it is verified.
-        if (submissionIdx > 0) {
-            uint16 verified = Uint16VectorLib.getUint16(
-                verifierStorage.numVerifiedInSubmission,
-                submissionIdx
-            );
-            if (verified == numProofs) {
-                return true;
+                // Assume all Submissions with the same SubmissionIdx have the
```
Comment (dysfunctor) [inline upa/contracts/UpaVerifier.sol]:
```suggestion
                // Assume all Submissions with the same SubmissionId have the
```

File: upa/scripts/test_upa [hunk 4]
```diff
@@ -346,19 +336,25 @@ pushd _test_upa
...
-    upa compute submission-proof \
-           --submission-file submit-multi-4.submission.json \
-           --start-idx 1 \
-           --num-proofs 3 \
-           > submit-multi-4.submission-proof.2.3.4.json
+    # upa compute submission-proof \
+    #        --submission-file submit-multi-4.submission.json \
+    #        --start-idx 1 \
+    #        --num-proofs 3 \
+    #        > submit-multi-4.submission-proof.2.3.4.json
```
Comment (dysfunctor) [inline upa/scripts/test_upa]: old code

File: upa/scripts/test_upa [hunk 5]
```diff
@@ -346,19 +336,25 @@ pushd _test_upa
...
+           --offset 1 \
+           --off-chain-submission off-chain-submission.json \
+           --off-chain-offset 4 \
            --wait > agg-proof.hash
 
+           # --submission-proof-file submit-multi-4.submission-proof.2.3.4.json \
+           # --proof-ids-file agg.pids.json \
+           # --submission-markers-file off-chain-submission-markers-2.json \
+           # --wait > agg-proof.hash
+
```
Comment (dysfunctor) [inline upa/scripts/test_upa]: old code

File: upa/src/tool/aggregator.ts [hunk 6]
```diff
@@ -210,31 +171,55 @@ const submitAggregatedProof = command({
...
+      long: "final-count",
+      short: "f",
+      description:
+        "include only leading proofs from final final submission" +
+        " (default: all)",
+    }),
+    offChainSubmissionFiles: multioption({
+      type: array(string),
+      long: "off-chain-submission",
+      description: "on-chain submission files",
```
Comment (dysfunctor) [inline upa/src/tool/aggregator.ts]:
```suggestion
      description: "off-chain submission files",
```

File: upa/src/tool/aggregator.ts [hunk 7]
```diff
@@ -210,31 +171,55 @@ const submitAggregatedProof = command({
...
+      short: "",
+      defaultValue: () => 0,
+      description: "skip proofs in first on-chain submission (default: 0)",
+    }),
+    finalCount: option({
+      type: optional(number),
+      long: "final-count",
+      short: "f",
+      description:
+        "include only leading proofs from final final submission" +
```
Comment (dysfunctor) [inline upa/src/tool/aggregator.ts]:
```suggestion
        "include only leading proofs from final submission" +
```

File: upa/src/tool/aggregator.ts [hunk 8]
```diff
@@ -210,31 +171,55 @@ const submitAggregatedProof = command({
...
+    offChainSubmissionFiles: multioption({
+      type: array(string),
+      long: "off-chain-submission",
+      description: "on-chain submission files",
+    }),
+    offChainOffset: option({
+      type: number,
+      long: "off-chain-offset",
+      defaultValue: () => 0,
+      description: "skip proofs in first on-chain submission (default: 0)",
```
Comment (dysfunctor) [inline upa/src/tool/aggregator.ts]:
```suggestion
      description: "skip proofs in first off-chain submission (default: 0)",
```

File: upa/src/tool/aggregator.ts [hunk 9]
```diff
@@ -210,31 +171,55 @@ const submitAggregatedProof = command({
...
+    }),
+    offChainFinalCount: option({
+      type: optional(number),
+      long: "off-chain-final-count",
+      description:
+        "include only leading proofs from final final submission" +
```
Comment (dysfunctor) [inline upa/src/tool/aggregator.ts]:
```suggestion
        "include only leading proofs from final submission" +
```

File: upa/src/sdk/client.ts [hunk 10]
```diff
@@ -56,13 +56,28 @@ export class UpaClient {
...
-    );
+    // console.log(
+    //   `SubmissionId ${submission.getSubmissionId()} ` +
+    //   `(txid: ${txResponse.hash})`
+    // );
 
     return { submission, txResponse };
   }
 
+  /// Wait for the submission to be successfully sent to the contract, and
```
Comment (dysfunctor) [inline upa/src/sdk/client.ts:67]: unfinished comment

File: upa/scripts/test_upa [hunk 11]
```diff
@@ -346,19 +336,25 @@ pushd _test_upa
...
-           > submit-multi-4.submission-proof.2.3.4.json
+    # upa compute submission-proof \
+    #        --submission-file submit-multi-4.submission.json \
+    #        --start-idx 1 \
+    #        --num-proofs 3 \
+    #        > submit-multi-4.submission-proof.2.3.4.json
     KEYFILE_PASSWORD=${DUMMY_PW} upa aggregator submit-aggregated-proof \
            --keyfile ${KEYFILE} \
            --calldata-file agg.proof.calldata \
-           --submission-proof-file submit-multi-4.submission-proof.2.3.4.json \
```
Comment (dysfunctor) [inline upa/scripts/test_upa:360]: great not to need these any more

File: upa/test/censorshipClaimTest.ts [hunk 12]
```diff
@@ -138,17 +145,22 @@ describe("Censorship challenge tests", () => {
...
-        packOffChainSubmissionMarkers([])
-      );
+    const submissionInterval = siFromSubmission(s, undefined);
+    const aggProofParams = computeAggregatedProofParameters(
+      [submissionInterval],
+      []
+    );
+
+    // const proofIds = s.proofIds;
+    // const submissionProof = s.computeSubmissionProof(0, proofIds.length);
```
Comment (dysfunctor) [inline upa/test/censorshipClaimTest.ts]: old code

File: upa/test/censorshipClaimTest.ts [hunk 13]
```diff
@@ -223,6 +236,65 @@ describe("Censorship challenge tests", () => {
...
+    await verifySubmission(verifier, worker, s3);
+    await verifySubmission(verifier, worker, dupSubmission);
+
+    expect(
+      await verifier.getFunction(isSubmissionVerifiedById)(s2.submissionId)
+    ).is.true;
+
+    const numProofsInS2 = s2.circuitIds.length;
+
+    // Challenege to upd shoudl fail
```
Comment (dysfunctor) [inline upa/test/censorshipClaimTest.ts]:
```suggestion
    // Challenge to upa should fail
```

**Comment by dysfunctor** [comment]
Gas numbers look quite good, I was expecting the impact to be bigger.

File: upa/src/tool/aggregator.ts (hunk 7 above)
Comment (dysfunctor) [reply upa/src/tool/aggregator.ts]: I wasn't sure what this meant from just the description text

File: upa/contracts/UpaVerifier.sol (hunk 2 above)
Comment (dtebbs) [reply upa/contracts/UpaVerifier.sol]: Thanks.  Good catch.  Yes, I can't think of a reason not to, now.  All tests pass so I'll do it that way.

File: spec/protocol/upa-1.2.0.md (hunk 1 above)
Comment (dtebbs) [reply spec/protocol/upa-1.2.0.md]: Fixed.  THanks.

File: upa/contracts/UpaVerifier.sol (hunk 3 above)
Comment (dtebbs) [reply upa/contracts/UpaVerifier.sol]: Fixed.  Thanks.

File: upa/scripts/test_upa (hunk 4 above)
Comment (dtebbs) [reply upa/scripts/test_upa]: removed

File: upa/scripts/test_upa (hunk 5 above)
Comment (dtebbs) [reply upa/scripts/test_upa]: removed

File: upa/src/tool/aggregator.ts (hunk 6 above)
Comment (dtebbs) [reply upa/src/tool/aggregator.ts]: updated

File: upa/src/tool/aggregator.ts (hunk 7 above)
Comment (dtebbs) [reply upa/src/tool/aggregator.ts]: fixed.
It's for when the last submission in the list is only partially included in the aggregated batch.  Any suggestions for better wording?

File: upa/src/tool/aggregator.ts (hunk 8 above)
Comment (dtebbs) [reply upa/src/tool/aggregator.ts]: Fixed

File: upa/src/sdk/client.ts (hunk 10 above)
Comment (dtebbs) [reply upa/src/sdk/client.ts:67]: Finished.  Thanks.

File: upa/test/censorshipClaimTest.ts (hunk 12 above)
Comment (dtebbs) [reply upa/test/censorshipClaimTest.ts]: removed.

File: upa/test/censorshipClaimTest.ts (hunk 13 above)
Comment (dtebbs) [reply upa/test/censorshipClaimTest.ts]: Fixed.  Thanks.

**Review by dysfunctor — APPROVED** [review]

**Pull request merged**